File: sdk-android/src/main/java/cu/suitetecsa/sdk/android/utils/SimCardUtils.java
package cu.suitetecsa.sdk.android.utils;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.telephony.TelephonyManager;

import androidx.annotation.RequiresApi;
import androidx.annotation.RequiresPermission;

import cu.suitetecsa.sdk.android.balance.FetchBalanceCallBack;
import cu.suitetecsa.sdk.android.balance.RequestCallback;
import cu.suitetecsa.sdk.android.balance.UssdRequestSender;
import cu.suitetecsa.sdk.android.balance.consult.UssdRequest;
import cu.suitetecsa.sdk.android.balance.response.UssdResponse;
import cu.suitetecsa.sdk.android.model.SimCard;

public class SimCardUtils {
    public static void makeCall(SimCard simCard, Context context, String phoneNumber) {
        Intent intent = new Intent(Intent.ACTION_CALL).setData(Uri.parse("tel:" + phoneNumber));

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            intent.putExtra("android.telecom.extra.PHONE_ACCOUNT_HANDLE", String.valueOf(simCard.subscriptionId()));
        } else {
            intent.putExtra("con.android.phone.extra.PHONE_ACCOUNT_HANDLE", String.valueOf(simCard.subscriptionId()));
        }

        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(intent);
    }

    @RequiresApi(Build.VERSION_CODES.O)
    @RequiresPermission(android.Manifest.permission.CALL_PHONE)
    public static void smartFetchBalance(SimCard simCard, FetchBalanceCallBack callBack) {
        RequestCallback requestCallback = new RequestCallback() {
            @Override
            public TelephonyManager getTelephonyManager() {
                return simCard.telephony();
            }

            @Override
            public void onRequesting(UssdRequest request) {
                callBack.onFetching(request);
            }

            @Override
            public void onSuccess(UssdRequest request, UssdResponse response) {
                callBack.onSuccess(request, response);
            }

            @Override
            public void onFailure(Throwable throwable) {
                callBack.onFailure(throwable);
            }
        };
        new UssdRequestSender.Builder().build().send(requestCallback);
    }

    @RequiresApi(Build.VERSION_CODES.O)
    @RequiresPermission(android.Manifest.permission.CALL_PHONE)
    public static void ussdFetch(SimCard simCard, String ussdCode, FetchBalanceCallBack callBack) {
        RequestCallback requestCallback = new RequestCallback() {
            @Override
            public TelephonyManager getTelephonyManager() {
                return simCard.telephony();
            }

            @Override
            public void onRequesting(UssdRequest request) {
                callBack.onFetching(request);
            }

            @Override
            public void onSuccess(UssdRequest request, UssdResponse response) {
                callBack.onSuccess(request, response);
            }

            @Override
            public void onFailure(Throwable throwable) {
                callBack.onFailure(throwable);
            }
        };
        new UssdRequestSender.Builder().build().send(ussdCode, requestCallback);
    }
}

File: sdk-android/src/main/java/io/github/suitetecsa/sdk/android/utils/SimCardUtils.kt
package io.github.suitetecsa.sdk.android.utils

import android.Manifest
import android.content.Context
import android.content.Intent
import android.telephony.TelephonyManager
import androidx.annotation.RequiresPermission
import androidx.core.net.toUri
import io.github.suitetecsa.sdk.android.balance.RequestCallback
import io.github.suitetecsa.sdk.android.balance.UssdRequestSender
import io.github.suitetecsa.sdk.android.balance.UssdStringCallback
import io.github.suitetecsa.sdk.android.model.SimCard

object SimCardUtils {
    @JvmStatic
    @RequiresPermission(Manifest.permission.CALL_PHONE)
    fun ussdFetch(simCard: SimCard, ussdCode: String, callBack: UssdStringCallback) {
        val requestCallback: RequestCallback = object : RequestCallback {
            override val telephonyManager: TelephonyManager
                get() = simCard.telephony

            override fun onSuccess(response: String) {
                callBack.onSuccess(response)
            }

            override fun onFailure(throwable: Throwable) = callBack.onFailure(throwable)
        }
        UssdRequestSender.Builder().build().send(ussdCode, requestCallback)
    }
}

@RequiresPermission(Manifest.permission.CALL_PHONE)
fun SimCard.ussdFetch(ussdCode: String, callBack: UssdStringCallback) =
    SimCardUtils.ussdFetch(this, ussdCode, callBack)
